mlflow/pyfunc/loaders/chat_agent.py

```diff
@@ -0,0 +1,111 @@
+from typing import Any, Generator, Optional
+
+import pydantic
+
+from mlflow.exceptions import MlflowException
+from mlflow.models.utils import _convert_llm_ndarray_to_list
+from mlflow.protos.databricks_pb2 import INTERNAL_ERROR
+from mlflow.pyfunc.model import (
+    _load_context_model_and_signature,
+)
+from mlflow.types.llm import (
+    ChatAgentMessage,
+    ChatAgentParams,
+    ChatAgentResponse,
+)
+from mlflow.types.type_hints import model_validate
+from mlflow.utils.annotations import experimental
+
+
+def _load_pyfunc(model_path: str, model_config: Optional[dict[str, Any]] = None):
+    _, chat_agent, signature = _load_context_model_and_signature(model_path, model_config)
+    return _ChatAgentPyfuncWrapper(chat_agent=chat_agent, signature=signature)
+
+
+@experimental
+class _ChatAgentPyfuncWrapper:
+    """
+    Wrapper class that converts dict inputs to pydantic objects accepted by :class:`~ChatAgent`.
+    """
+
+    def __init__(self, chat_agent, signature):
+        """
+        Args:
+            chat_agent: An instance of a subclass of :class:`~ChatAgent`.
+            signature: :class:`~ModelSignature` instance describing model input and output.
+        """
+        self.chat_agent = chat_agent
+        self.signature = signature
+
+    def get_raw_model(self):
+        """
+        Returns the underlying model.
+        """
+        return self.chat_agent
+
+    def _convert_input(self, model_input):
+        import pandas
+
+        if isinstance(model_input, dict):
+            dict_input = model_input
+        elif isinstance(model_input, pandas.DataFrame):
+            dict_input = {
+                k: _convert_llm_ndarray_to_list(v[0])
+                for k, v in model_input.to_dict(orient="list").items()
+            }
+        else:
+            raise MlflowException(
+                "Unsupported model input type. Expected a dict or pandas.DataFrame, "
+                f"but got {type(model_input)} instead.",
+                error_code=INTERNAL_ERROR,
+            )
+
+        messages = [ChatAgentMessage(**message) for message in dict_input.pop("messages", [])]
+        params = ChatAgentParams(**dict_input)
+        return messages, params
+
+    # used for both streaming and non streaming
+    def _response_to_dict(self, response: ChatAgentResponse) -> dict[str, Any]:
+        try:
+            model_validate(ChatAgentResponse, response)
+        except pydantic.ValidationError as e:
+            raise MlflowException(
+                message=(
+                    "Model returned an invalid response. Expected a ChatAgentResponse object "
+                    "or dictionary with the same schema"
+                ),
+                error_code=INTERNAL_ERROR,
+            ) from e
```

Comment:
Let's include e in the message, as this makes the real exception hard to find :)